app/src/main/kotlin/com/wire/android/ui/home/settings/backup/BackupAndRestoreViewModel.kt

```diff
@@ -1,17 +1,272 @@
+            when (val result = importBackup(latestExtractedBackupRootPath, restorePassword.text)) {
+                RestoreBackupResult.Success -> {
+                    state = state.copy(
+                        backupRestoreProgress = BackupRestoreProgress.Finished,
+                        restorePasswordValidation = PasswordValidation.Valid
+                    )
+                }
+                is RestoreBackupResult.Failure -> {
+                    mapBackupRestoreFailure(result.failure)
+                }
+            }
+        } else {
+            state = state.copy(backupRestoreProgress = BackupRestoreProgress.Failed)
+        }
+    }
+
+    private fun mapBackupRestoreFailure(failure: RestoreBackupResult.BackupRestoreFailure) = when (failure) {
+        InvalidPassword -> state = state.copy(
+            backupRestoreProgress = BackupRestoreProgress.Failed,
+            restorePasswordValidation = PasswordValidation.NotValid,
+        )
+
+        InvalidUserId -> state = state.copy(
+            backupRestoreProgress = BackupRestoreProgress.Failed,
+            restoreFileValidation = RestoreFileValidation.WrongBackup,
+            restorePasswordValidation = PasswordValidation.Valid
+        )
+
+        is IncompatibleBackup -> state = state.copy(
+            backupRestoreProgress = BackupRestoreProgress.Failed,
+            restoreFileValidation = RestoreFileValidation.IncompatibleBackup,
+            restorePasswordValidation = PasswordValidation.Valid
+        )
+
+        is BackupIOFailure, is DecryptionFailure -> state = state.copy(
+            backupRestoreProgress = BackupRestoreProgress.Failed,
+            restoreFileValidation = RestoreFileValidation.GeneralFailure,
+            restorePasswordValidation = PasswordValidation.Valid
+        )
+    }
+
+    fun validateBackupCreationPassword(backupPassword: TextFieldValue) {
+        // TODO: modify in case the password requirements change
+    }
+
+    fun cancelBackupCreation() {
+        state = state.copy(
+            backupCreationProgress = BackupCreationProgress.Pending,
+        )
+    }
+
+    fun cancelBackupRestore() {
+        state = state.copy(
+            restoreFileValidation = RestoreFileValidation.Pending,
+            backupRestoreProgress = BackupRestoreProgress.Pending,
+            restorePasswordValidation = PasswordValidation.NotVerified
+        )
+    }
+
+    fun navigateToConversations() {
+        viewModelScope.launch {
+            navigationManager.navigate(NavigationCommand(NavigationItem.Home.getRouteWithArgs(), BackStackMode.CLEAR_WHOLE))
+        }
+    }
 
     fun navigateBack() = viewModelScope.launch { navigationManager.navigateBack() }
-    fun backupCreate() { /* TODO */ }
-    fun backupRestore() { /* TODO */ }
+
+    private companion object {
+        const val TEMP_IMPORTED_BACKUP_FILE_NAME = "tempImportedBackup.zip"
+    }
+}
+
+data class BackupAndRestoreState(
```

Comment:
I think making the file smaller would be a good thing, maybe we can extract this `BackupAndRestoreState` to a separete file